### ISSUE TITLE
document provenance requirements for new/existing code

### ISSUE THREAD

**Issue by petrelharp** [tskit-dev/msprime#968]
With the new provenance code there are some additional requirements for adding new methods or modifying existing ones, as I'm discovering over in #946. In figuring this out it'd be very helpful to have a short explainer and/or checklist of what needs to happen to interface with the provenance machinery. For instance: an `.asdict()` method... ? Right now I'm going on "what do errors in running tests tell me", and I'm starting to get a sense for how things work (which seems very nice!), but the big picture would really help. Apologies if this exists somewhere and I have missed it.

**Comment by benjeffery** [comment]
Agreed! I should have done this as part of the initial work. I'll self-assign this.

**Comment by petrelharp** [comment]
Great! Thanks.

**Comment by jeromekelleher** [comment]
I'm going to drop this out of the 1.0 milestone because it's more of an internal development thing and not really user facing.
